### PR TITLE
feat(parser): support regex matching on JSON arrays with scalar elements

### LINKED ISSUE #1157: Git-cliff does not parse changes after breaking change (tag sha-4c3c946)
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Description of the bug

When I run `git cliff` (v. 2.8.0) from Powershell against my repo, it correctly populates the changelog. If I use a Docker image with the tag `sha-4c3c946` or prior, it populates the changelog correctly. However, when I try to use the latest Docker image, or any image after tag `sha-4c3c946`, it creates an empty changelog with only a header and footer. 

I compiled the latest version of git-cliff (3 days prior to v2.9.0 release) and came up with the same results where it produces a nearly empty changelog.

It appears to be an issue related to [issue 1143](https://github.com/orhun/git-cliff/issues/1143), which resulted in [this commit](https://github.com/orhun/git-cliff/commit/e981e1d1b27a65fc2d2fd51b025c27692a6c8910).

### Steps To Reproduce

1. Copy down the [mta-solutions/fsharp-data-validation](https://github.com/[mta-solutions/fsharp-data-validation](https://github.com/mta-solutions/fsharp-data-validation)) GitHub Repo (this is just for illustrative purposes) to `D:/fsharp-data-validation/` (or the directory of your choice). 
2. Create a `cliff.toml` file in the root of the repo with the following contents:
   ```toml
   [remote.github]
   owner = "mta-solutions"
   repo = "fsharp-data-validation"
   
   [changelog]
   output = "CHANGELOG.md"
   
   header = """
   # Changelog
   
   """
   
   body = """
   {% if version %}\
       ## [{{ version | trim_start_matches(pat="v") }}](https://github.com/{{ remote.github.owner }}/{{ remote.github.repo }}/tree/{{ version }}) ({{ timestamp | date(format="%Y-%m-%d") }})
   {% else %}\
       {% if commits | length > 0 %}\
           \n## [unreleased]
       {% endif %}\
   {% endif %}\
   
   {% for group, commits in commits | unique(attribute="id") | filter(attribute="merge_commit", value=true) | group_by(attribute="group") %}
       ### {{ group | upper_first }}
       {% for commit in commits %}
           - {{ commit.message | split(pat="\n") | nth(n=2) | trim_end }} \
           ([PR #{{ commit.remote.pr_number }}](https://github.com/{{ remote.github.owner }}/{{ remote.github.repo }}/pull/{{ commit.remote.pr_number }})) \
           ([{{ commit.remote.username }}](https://github.com/{{ commit.remote.username }}))\
       {% endfor %}
   {% endfor %}\n
   """
   
   footer = """
   \n\n*This changelog was automatically generated by [git-cliff](https://github.com/orhun/git-cliff)*
   """
   
   trim = true
   
   [git]
   conventional_commits = false
   filter_unconventional = false
   split_commits = false
   commit_parsers = [
     { field = "remote.pr_labels", pattern = "duplicate|invalid|wontfix|skip changelog", skip = true },
     { field = "remote.pr_labels", pattern = "breaking change", group = "<!-- 0 -->🏗️ Breaking Changes" },
     { field = "remote.pr_labels", pattern = "feature|deprecation", group = "<!-- 1 -->🚀 Features" },
     { field = "remote.pr_labels", pattern = "enhancement|refactor", group = "<!-- 1 -->🛠️ Enhancements" },
     { field = "remote.pr_labels", pattern = "bug|regression", group = "<!-- 2 -->🐛 Bug Fixes" },
     { field = "remote.pr_labels", pattern = "security", group = "<!-- 3 -->🔐 Security" },
     { field = "remote.pr_labels", pattern = "documentation", group = "<!-- 4 -->📝 Documentation" },
     { field = "remote.pr_labels", pattern = ".*", group = "<!-- 5 -->🌀 Miscellaneous" },
   ]
   filter_commits = true
   sort_commits = "newest"
   ```
3. Run this command from a Powershell console (adjust the repo directory if necessary): `docker run -it --rm --env GITHUB_TOKEN -v D:/fsharp-data-validation:/app orhunp/git-cliff:latest`
4. View the `CHANGELOG.md`, which should be empty except for a single tag and the header and footer.

### Expected behavior

By following the same directions to reproduce, but replacing the command in step three with `docker run -it --rm --env GITHUB_TOKEN -v D:/fsharp-data-validation:/app orhunp/git-cliff:sha-4d3b2d5` (using `sha-4d3b2d5` any prior tag), the `CHANGELOG.md` file will be fully populated.

### Screenshots / Logs

I manually compiled the latest git-cliff v2.8.0 (commit 7148b2dbff518eb71c276f3c256db49970730863) and ran it with the `--verbose` option against the sample repo and this is what I received:

```
D:\git-cliff\target\release\git-cliff.exe --verbose
 INFO  git_cliff > A new version of git-cliff is available: v2.8.0 -> v2.9.0
 DEBUG git_cliff_core::changelog > Adding remote data...
 DEBUG git_cliff_core::changelog > You are using an experimental feature! Please report bugs at <https://git-cliff.org/issues>
▸▹▹▹▹ Retrieving data from GitHub... (mta-solutions/fsharp-data-validation)                                                                                                DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=0
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=1
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=2
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=3
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=4
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=5
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=6
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=7
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=8
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/commits?per_page=100&page=9
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/pulls?per_page=100&page=0&state=closed
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/pulls?per_page=100&page=1&state=closed
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/pulls?per_page=100&page=2&state=closed
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/pulls?per_page=100&page=3&state=closed
 DEBUG git_cliff_core::remote    > Sending request to: https://api.github.com/repos/mta-solutions/fsharp-data-validation/pulls?per_page=100&page=4&state=closed
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG reqwest::connect          > starting new connection: https://api.github.com/
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
 DEBUG rustls::client::hs        > No cached session for DnsName("api.github.com")
 DEBUG rustls::client::hs        > Not resuming any session
▹▸▹▹▹ Retrieving data from GitHub... (mta-solutions/fsharp-data-validation)                                                                                                DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
 DEBUG rustls::client::hs        > Using ciphersuite TLS13_AES_128_GCM_SHA256
 DEBUG rustls::client::tls13     > Not resuming
 DEBUG rustls::client::tls13     > TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
 DEBUG rustls::client::hs        > ALPN protocol is Some(b"http/1.1")
▸▹▹▹▹ Retrieving data from GitHub... (mta-solutions/fsharp-data-validation)                                                                                                DEBUG git_cliff_core::remote    > Error while fetching page: PaginationError("end of entries")
▹▸▹▹▹ Retrieving data from GitHub... (mta-solutions/fsharp-data-validation)                                                                                                DEBUG git_cliff_core::remote    > Error while fetching page: PaginationError("end of entries")
 DEBUG git_cliff_core::changelog > Number of GitHub commits: 168
 DEBUG git_cliff_core::changelog > Number of GitHub pull requests: 54
 DEBUG git_cliff_core::changelog > Processing the commits...
 DEBUG git_cliff_core::changelog > Processing 8 release(s)...
 DEBUG git_cliff_core::changelog > Generating changelog...
```

I also ran it with the tagged v2.8.0 release version, which came up with the same results.

When I run git cliff with the context, the context for the working/older versions of git cliff produce a mile-long context. When I do the same with the compiled version, it produces a tiny context that includes barely anything.

### Software information

- Operating system: Windows 11 Pro
- Rust version: 1.85.0
- Project version: 2.8.0/latest

### Additional context

When I remove the `filter_commits` and `commit_parsers` options from the config, then it produces a large context and changelog, though things are obviously not categorized or anything correctly then. This makes me suspect something changed with the commit_parses.

I have also tried this with at least three different private repos where we have every kind of label and it is the same issue.

### PR DESCRIPTION
**Pull request by ognis1205**
## Description

This PR extends the commit parser to support regex matching on JSON arrays, where each element is a scalar value (e.g., strings, numbers, booleans). When a commit field (like `remote.pr_labels`) is a JSON array, all entries are evaluated according to the specified `match_mode` (`any` or `all`). Fixes #1157

## Motivation and Context

Previously, only scalar JSON values (e.g., strings, numbers) were matched against regex patterns. This change extends support to JSON arrays, where each element is a scalar value. All values in the array are checked based on the configured match mode (`any`/`all`).

## How Has This Been Tested?

 1. Clone the [mta-solutions/fsharp-data-validation](https://github.com/mta-solutions/fsharp-data-validation) GitHub Repo.
 2. Create a cliff.toml file in the root of the repo with the following contents:

```toml
[remote.github]
owner = "mta-solutions"
repo = "fsharp-data-validation"

[changelog]
output = "CHANGELOG.md"

header = """
# Changelog

"""

body = """
{% if version %}\
    ## [{{ version | trim_start_matches(pat="v") }}](https://github.com/{{ remote.github.owner }}/{{ remote.github.repo }}/tree/{{ version }}) ({{ timestamp | date(format="%Y-%m-%d") }})
{% else %}\
    {% if commits | length > 0 %}\
        \n## [unreleased]
    {% endif %}\
{% endif %}\

{% for group, commits in commits | unique(attribute="id") | filter(attribute="merge_commit", value=true) | group_by(attribute="group") %}
    ### {{ group | upper_first }}
    {% for commit in commits %}
        - {{ commit.message | split(pat="\n") | nth(n=2) | trim_end }} \
        ([PR #{{ commit.remote.pr_number }}](https://github.com/{{ remote.github.owner }}/{{ remote.github.repo }}/pull/{{ commit.remote.pr_number }})) \
        ([{{ commit.remote.username }}](https://github.com/{{ commit.remote.username }}))\
    {% endfor %}
{% endfor %}\n
"""

footer = """
\n\n*This changelog was automatically generated by [git-cliff](https://github.com/orhun/git-cliff)*
"""

trim = true

[git]
conventional_commits = false
filter_unconventional = false
split_commits = false
commit_parsers = [
  { field = "remote.pr_labels", pattern = "duplicate|invalid|wontfix|skip changelog", skip = true },
  { field = "remote.pr_labels", pattern = "breaking change", group = "<!-- 0 -->🏗️ Breaking Changes" },
  { field = "remote.pr_labels", pattern = "feature|deprecation", group = "<!-- 1 -->🚀 Features" },
  { field = "remote.pr_labels", pattern = "enhancement|refactor", group = "<!-- 1 -->🛠️ Enhancements" },
  { field = "remote.pr_labels", pattern = "bug|regression", group = "<!-- 2 -->🐛 Bug Fixes" },
  { field = "remote.pr_labels", pattern = "security", group = "<!-- 3 -->🔐 Security" },
  { field = "remote.pr_labels", pattern = "documentation", group = "<!-- 4 -->📝 Documentation" },
  { field = "remote.pr_labels", pattern = ".*", group = "<!-- 5 -->🌀 Miscellaneous" },
]
filter_commits = true
sort_commits = "newest"
```

 3. Run this command from a console:
 
 ```bash
 path/to/my/local/git-cliff/repository/target/debug/git-cliff -v 
 ```
 
  4. View the `CHANGELOG.md`:
  
  ```md
  # Changelog

## [2.2.0](https://github.com/mta-solutions/fsharp-data-validation/tree/v2.2.0) (2024-12-13)

### <!-- 1 -->🛠️ Enhancements

- Add additional VCtx functions ([PR #24](https://github.com/mta-solutions/fsharp-data-validation/pull/24)) ([RJSonnenberg](https://github.com/RJSonnenberg))

## [2.1.0](https://github.com/mta-solutions/fsharp-data-validation/tree/v2.1.0) (2024-12-10)

### <!-- 1 -->🛠️ Enhancements

- Remove internal scoping and add Async extension methods for validation contexts ([PR #22](https://github.com/mta-solutions/fsharp-data-validation/pull/22)) ([RJSonnenberg](https://github.com/RJSonnenberg))

### <!-- 4 -->📝 Documentation

- Prep v2.1.0 Release ([PR #23](https://github.com/mta-solutions/fsharp-data-validation/pull/23)) ([RJSonnenberg](https://github.com/RJSonnenberg))

## [2.0.0](https://github.com/mta-solutions/fsharp-data-validation/tree/v2.0.0) (2024-10-08)

### <!-- 1 -->🛠️ Enhancements

- Use FSharpPlus ([PR #20](https://github.com/mta-solutions/fsharp-data-validation/pull/20)) ([devinlyons](https://github.com/devinlyons))

### <!-- 4 -->📝 Documentation

- Prep for Release v2.0.0 ([PR #21](https://github.com/mta-solutions/fsharp-data-validation/pull/21)) ([devinlyons](https://github.com/devinlyons))

## [1.0.1](https://github.com/mta-solutions/fsharp-data-validation/tree/v1.0.1) (2023-11-17)

### <!-- 1 -->🛠️ Enhancements

- Convert to NuGet ([PR #18](https://github.com/mta-solutions/fsharp-data-validation/pull/18)) ([devinlyons](https://github.com/devinlyons))
- Add Giraffe model validation library and examples ([PR #17](https://github.com/mta-solutions/fsharp-data-validation/pull/17)) ([RJSonnenberg](https://github.com/RJSonnenberg))

## [1.0.0](https://github.com/mta-solutions/fsharp-data-validation/tree/v1.0.0) (2022-03-14)

### <!-- 1 -->🛠️ Enhancements

- Add utility functions to help with using NonEmptyLists [MBI-53] ([PR #14](https://github.com/mta-solutions/fsharp-data-validation/pull/14)) ([mleone11](https://github.com/mleone11))



*This changelog was automatically generated by [git-cliff](https://github.com/orhun/git-cliff)*
  ```

## Screenshots / Logs (if applicable)

N/A

## Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ognis1205** [comment]
cc: @RJSonnenberg

File: git-cliff-core/src/commit.rs [hunk 1]
```diff
@@ -312,34 +313,49 @@ impl Commit<'_> {
...
+								Some(vec![v.to_string()])
+							}
+							Value::Array(arr) => {
+								let mut result = Vec::new();
+								for item in arr {
+									match item {
+										Value::String(s) => result.push(s.clone()),
+										Value::Number(_) |
+										Value::Bool(_) |
+										Value::Null => result.push(item.to_string()),
```
Comment (ognis1205) [inline git-cliff-core/src/commit.rs]: This line of the newly implemented code is missing from the code coverage, but since it is theoretically unreachable, I have deemed the lack of coverage here to be acceptable.

**Comment by ognis1205** [comment]
Hey @RJSonnenberg (cc: @orhun ) ,

I generated the `CHANGELOG.md` on my local working branch build using the repository you provided (for details, please refer to the “How Has This Been Tested?” section of the PR). Grouping by `remote.pr_labels` is working, but there's one difference in behavior compared to the existing setup.

```
{ field = "remote.pr_labels", pattern = ".*", group = "<!-- 5 -->🌀 Miscellaneous" },
```

Previously, I believe this served as the fallback group setting. However, with the default `MatchMode::Any`, this does not match when `remote.pr_labels` is empty. Instead, using:

```
{ message = ".*", group = "<!-- 6 --> 🌀 Miscellaneous" },
```

continues to work as a fallback as before.

Note: The original rule ends up matching an empty JSON array (`[]`) as a string, which is probably not the intended behavior. It might be worth adding a comment or warning about this to avoid confusion.

Any thoughts on this?

**Comment by RJSonnenberg** [comment]
> Hey @RJSonnenberg (cc: @orhun ) ,
> 
> I generated the `CHANGELOG.md` on my local working branch build using the repository you provided (for details, please refer to the “How Has This Been Tested?” section of the PR). Grouping by `remote.pr_labels` is working, but there's one difference in behavior compared to the existing setup.
> 
> ```
> { field = "remote.pr_labels", pattern = ".*", group = "🌀 Miscellaneous" },
> ```
> 
> Previously, I believe this served as the fallback group setting. However, with the default `MatchMode::Any`, this does not match when `remote.pr_labels` is empty. Instead, using:
> 
> ```
> { message = ".*", group = " 🌀 Miscellaneous" },
> ```
> 
> continues to work as a fallback as before.
> 
> Note: The original rule ends up matching an empty JSON array (`[]`) as a string, which is probably not the intended behavior. It might be worth adding a comment or warning about this to avoid confusion.
> 
> Any thoughts on this?

I think that is a fine compromise. Like you said, the original only worked as a fallback, so the new fallback group setting looks fine to me. It might throw people off, but as long as it is documented well, we should be good.

**Comment by ognis1205** [comment]
@RJSonnenberg 

Thanks for your feedback!

File: git-cliff-core/src/config.rs [hunk 2]
```diff
@@ -363,6 +374,9 @@ pub struct CommitParser {
 	/// Regex for matching the field value.
 	#[serde(with = "serde_regex", default)]
 	pub pattern:       Option<Regex>,
+	/// Mode to match regex against fields: any or all.
+	#[serde(default)]
+	pub match_mode:    MatchMode,
```
Comment (orhun) [inline git-cliff-core/src/config.rs]: From reading the changes, it's not clear to me why this was added to fix this problem. Can you share some details?

Also, we should document this if it is a necessary change :)

**Comment by orhun** [comment]
Hey guys! I just had some time catch up with #1157 and this PR - thanks for investigating and possible fix!

I'd like to take a step back and see a minimal example what's broken for understanding this fix better. @ognis1205 can you add a new [fixture test](https://github.com/orhun/git-cliff/tree/main/.github/fixtures)? Preferably for grouping `pr_labels` which is know allegedly broken. I also think that adding that fixture will also help us to catch these regressions better in the future.

File: git-cliff-core/src/config.rs (hunk 2 above)
Comment (ognis1205) [reply git-cliff-core/src/config.rs]: Thanks for the question!

To clarify with a concrete case: imagine we want to match PRs where at least one of the labels matches a pattern like `"enhancement|refactor"`. If a commit  has `remote.pr_labels = ["feature", "refactor"]`, then:

 - With `match_mode = "any"`, this is a match (since `"refactor"` matches).
 - With `match_mode = "all"`, it's not a match (since `"feature"` doesn't match).
 
 ###  Detailed Example
 
 ```
 commit_parsers = [
  { field = "remote.pr_labels", pattern = "enhancement|refactor", group = "<!-- 1 -->🛠️ Enhancements" },
]
 ```
 
 Given:
 
 ```
 remote.pr_labels = ["feature", "refactor"]
 ```
 
 ✅ MatchMode::Any

 - "feature" → ❌ (does not match)
 - "refactor" → ✅ (matches)

→ Since at least one label matches, the whole condition is satisfied.

❌ MatchMode::All

 - "feature" → ❌
 - "refactor" → ✅

→ Since not all labels match, the match fails.

### Explanation

Previously, `CommitParser` implicitly treated the `remote.pr_labels` field as if `MatchMode::Any` were applied, except when the field was empty.
With this internal changes where the `field` is now handled as a `Vec`, I introduced `match_mode` to make the aggregation behavior of regex matching across vector elements explicit at the API level — i.e., whether the matches should be interpreted as an `AND` or an `OR`.
This change ensures that the matching semantics are both predictable and configurable moving forward.

**Comment by ognis1205** [comment]
> Hey guys! I just had some time catch up with #1157 and this PR - thanks for investigating and possible fix!
> 
> I'd like to take a step back and see a minimal example what's broken for understanding this fix better. @ognis1205 can you add a new [fixture test](https://github.com/orhun/git-cliff/tree/main/.github/fixtures)? Preferably for grouping `pr_labels` which is know allegedly broken. I also think that adding that fixture will also help us to catch these regressions better in the future.

@orhun

Thanks for the feedback! You're absolutely right that having a fixture test would help with catching regressions like this.

However, there’s a limitation in our current fixture test infrastructure:
PR labels (`remote.pr_labels`) are metadata provided by remote Git services. Since, I believe, our fixture tests currently rely only on actual Git history via `git2::Repository` and `revwalk` (not `get_(github|gitlab|gitea|bitbucket)_metadata` called), there's no built-in mechanism to inject or simulate remote metadata like PR labels during commit collection.

To support this in fixture tests, we would need to extend the logic to programmatically attach synthetic `remote.pr_labels` to `Commit` objects after walking the Git history.

Any thoughts on this?

File: git-cliff-core/src/config.rs (hunk 2 above)
Comment (orhun) [reply git-cliff-core/src/config.rs]: Hmm I see... I still think that we can avoid this new field though. Maybe loop through each value in `Value::Array` and add them to the `regex_checks` array, wouldn't that work?

If we are going to go forward with it then we should document that this is useful for array types and also figure out if there are any other edge cases.

**Comment by orhun** [comment]
For the fixtures, you can create a dummy repository on GitHub and simply use that, like we do [here](https://github.com/orhun/git-cliff/blob/main/.github/fixtures/test-github-integration/commit.sh) or  [here](https://github.com/orhun/git-cliff/blob/main/.github/fixtures/test-gitlab-integration/commit.sh)

**Comment by ognis1205** [comment]
@orhun

> For the fixtures, you can create a dummy repository on GitHub and simply use that, like we do [here](https://github.com/orhun/git-cliff/blob/main/.github/fixtures/test-github-integration/commit.sh) or [here](https://github.com/orhun/git-cliff/blob/main/.github/fixtures/test-gitlab-integration/commit.sh)

Thanks! I will add a fixture test that uses a dummy GitHub repository as suggested.

File: git-cliff-core/src/config.rs (hunk 2 above)
Comment (ognis1205) [reply git-cliff-core/src/config.rs]: You're right, and I think the behavior you're describing is actually quite similar to what I implemented with `MatchMode::Any`. It maintains backward compatibility while enabling more flexible matching for array values.

I'll try rewriting the implementation with this in mind and see if it simplifies the design. Thanks for the input!

Comment (orhun) [reply git-cliff-core/src/config.rs]: Cool, let me know about your progress :)

Comment (ognis1205) [reply git-cliff-core/src/config.rs]: Thanks! I’ve implemented the changes as you suggested.
One thing to note—as I mentioned in [this comment](https://github.com/orhun/git-cliff/pull/1163#issuecomment-2947760486)—is that the following fallback:

```
{ field = "remote.pr_labels", pattern = ".*", group = "<!-- 5 -->🌀 Miscellaneous" },
```

no longer works as expected. Instead, you’ll need to use:

```
{ message = ".*", group = "<!-- 6 --> 🌀 Miscellaneous" },
```

The core logic update is complete, and the only remaining task is to add fixtures. I plan to address that as discussed in [this thread](https://github.com/orhun/git-cliff/discussions/1169).

Comment (ognis1205) [reply git-cliff-core/src/config.rs]: @orhun (cc: @RJSonnenberg)

Also, as an initial fixture test, I’ve added one based on the commit history from this repository:

https://github.com/mta-solutions/fsharp-data-validation

That said, if you think it’s better for maintainability to have a dedicated test case, as we discussed in the [discussion thread](https://github.com/orhun/git-cliff/discussions/1169), I can also add a commit history that uses PR labels explicitly for testing purposes. Let me know what you think :)

**Review by RJSonnenberg — APPROVED** [review]

**Review by orhun — APPROVED** [review]
LGTM, thanks!

File: git-cliff-core/src/config.rs (hunk 2 above)
Comment (orhun) [reply git-cliff-core/src/config.rs]: Nice!

We can update/expand the fixtures after that discussion is resolved :)

Thanks for your work on this!

**Comment by ognis1205** [comment]
@orhun 

I really appreciate your patience and thoroughness in the reviews. Thank you!

File: git-cliff-core/src/config.rs (hunk 2 above)
Comment (ognis1205) [reply git-cliff-core/src/config.rs]: Sure — we’ll continue the discussion there and figure out the best approach.  
Thanks again for your support!

**Comment by orhun** [comment]
Thanks for implementing!
